### DISCUSSION TITLE
line endings directive ignored?

### DISCUSSION THREAD

**Discussion by dpprdan** [twpayne/chezmoi#3816] [Q&A] [answered]
I am trying to add a [bitwarden notes field](https://bitwarden.com/help/cli/#get) to a template and use the [line endings directive](https://www.chezmoi.io/reference/templates/directives/#line-endings) so that the target file has native line endings (i.e. crlf on Windows).

My `config.tmpl` template looks like this

```go
chezmoi:template:line-endings=native
{{- (bitwarden "item" "my.note").notes -}}
```

I've [also tried](https://github.com/twpayne/chezmoi/discussions/2824#discussioncomment-5271290)

```go
{{- chezmoi:template:line-endings=native -}}
{{- (bitwarden "item" "my.note").notes -}}
```

However, on Windows, the `config` target file gets written with LF line endings, and not CRLF.

Am I doing something wrong or could this be a bug?

**Comment by halostatue** [comment]
This *might* be a bug, and it doesn't look like something that can be tested with `chezmoi execute-template` (there's no `--line-endings` parameter). Try using `line-endings=crlf` to see if it outputs the data from `.notes` with CRLF at all.

**Comment by bradenhilton** [comment]
Here is a short example template (`test.txt.tmpl`):

(I deleted `test.txt` before `chezmoi apply`, so chezmoi created the file each time.)

```text
chezmoi:template:line-endings=crlf
hello
world
foo
bar
```

- Source file line endings: LF
- Destination file line endings: LF:

    ```console
    ❯ Get-Content -Raw test.txt | Format-Hex
    
       Label: String (System.String) <51BAFB1D>
    
              Offset Bytes                                           Ascii
                     00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
              ------ ----------------------------------------------- -----
    0000000000000000 68 65 6C 6C 6F 0A 77 6F 72 6C 64 0A 66 6F 6F 0A hello�world�foo�
    0000000000000010 62 61 72 0A                                     bar�
    ```

---

```text
chezmoi:template:line-endings=lf
hello
world
foo
bar
```

- Source file line endings: CRLF
- Destination file line endings: CRLF:

    ```console
    ❯ Get-Content -Raw test.txt | Format-Hex
    
       Label: String (System.String) <3C87E777>
    
              Offset Bytes                                           Ascii
                     00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
              ------ ----------------------------------------------- -----
    0000000000000000 68 65 6C 6C 6F 0D 0A 77 6F 72 6C 64 0D 0A 66 6F hello��world��fo
    0000000000000010 6F 0D 0A 62 61 72 0D 0A                         o��bar��
    ```

I think it's reasonable to expect the destination file to have the line endings specified in the directive. I'm not sure at the moment if this was never the case or it's a regression, as I haven't looked into line endings in a while.

**Comment by twpayne** [comment, marked as answer]
Thanks for reporting this, this is actually a bug.

The documentation states that the directive is `chezmoi:template:line-endings=native` but the code is actually looking for `chezmoi:template:line-ending=native` (singular `ending`, not plural `endings`).

#3818 fixes this. In the short term, you can use

    chezmoi:template:line-ending=native

in your template.
